### ISSUE TITLE
Show all errors from cloudformation in CLI on stack update failure

### ISSUE THREAD

**Issue by DavidWells** [serverless/serverless#1818]
##### Serverless Framework Version: 
1.0.0-beta.1.1

##### Operating System: 
OSX Yosemite 10.10.5 (14F27)

##### Expected Behavior:

Show all errors that occurred to cause cloudformation to fail in CLI, not just the last one.

##### Actual Behavior:

![image](https://cloud.githubusercontent.com/assets/532272/17573503/9c062528-5f0f-11e6-8eb9-173e52852989.png)

This error is in the logs but the root cause was an error that caused the issue

![image](https://cloud.githubusercontent.com/assets/532272/17573574/27cf1e3e-5f10-11e6-8d3c-c7d215fc02cd.png)

Actual error was `Property AttributeDefinitions is inconsistent with the KeySchema of the table and the secondary indexes` but I had to go into AWS admin to see it.

**Comment by DavidWells** [comment]
Here is another example:

![image](https://cloud.githubusercontent.com/assets/532272/17573812/be28e8f0-5f11-11e6-9049-d6bbf3024c29.png)

![image](https://cloud.githubusercontent.com/assets/532272/17573840/f0a37408-5f11-11e6-8401-a4aa9cd27f0b.png)

**Comment by ac360** [comment]
I ran into this problem is well.  On my first deployment and many others.

This has been one of the biggest requested features for a long time (even in V.0).  As CloudFormation is central to V.1, I think this is a bigger priority than ever.  It's currently killing the core development experience.

Fortunately, this doesn't seem hard to fix.  Proposal:

* On failed deployment, call CloudFormation [`describeStackEvents`](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#describeStackEvents-property) via SDK.
* Loop through those events and pull out the ones that have a failed status.
* Either list all of the resources with a failed status and their reason OR list only the original one that failed, since others that fail after the first failed resource typically have the same error message.

**Comment by nicka** [comment]
Can't we just "tail" the CloudFormation deployment events during the deployment?

**Comment by ac360** [comment]
@nicka Interesting idea.  That would be more helpful because you would actually know what's going on :)

**Comment by danhumphrey** [comment]
+1 I had a similar issue. I made a mistake configuring a DynamoDB resource but it wasn't obvious why without logging into the AWS console.

**Comment by nicka** [comment]
We could fetch for new stack events every x seconds and break if any of the following states are noticed:

```
sucessStates = [
  "CREATE_COMPLETE",
  "UPDATE_COMPLETE"
]

failureStates = [
  "CREATE_FAILED",
  "DELETE_FAILED",
  "ROLLBACK_COMPLETE",
  "ROLLBACK_FAILED",
  "UPDATE_ROLLBACK_COMPLETE",
  "UPDATE_ROLLBACK_FAILED"
]
```

**Comment by ac360** [comment]
@nicka Exactly

**Comment by pmuens** [comment]
@nicka good idea! We actually already doing the monitoring here: https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/lib/updateStack.js#L39-L72

I would combine the ideas from @ac360 and @nicka so that we'll get the stack status and the detailed error message printed our altogether so that one has no need to sign in to the AWS console (been there way too often).

**Comment by ac360** [comment]
@pmuens Just ran into this issue again w/ first deployment using V.1 BETA.2.  It error'ed with no info.  2nd deployment error'ed again with the message `undefined`.  Definitely would like to see this in Beta.3

**Comment by pmuens** [comment]
@ac360 yes, just added it to the beta.3 milestone. Have you deployed a fresh new service? What was your CloudFormation error (was it Serverless related)?

**Comment by flomotlik** [comment]
Yes super important feature, this is a very obvious UX improvement for people. Should definitely be in Beta 3. If somebody picks it up even better, if not we're going to do it ourselves

**Comment by nicka** [comment]
@flomotlik I would be willing to take on this task.

**Comment by flomotlik** [comment]
Thats what I hoped :D. The only thing I worry about showing all events is that it will be a lot of noise, so one thing we could do is by default only show events that have failed and add a `--all-events` parameter that will show all events that are happening?

Most of the time if everything fails you're not really interested in it anyway I'd say.

**Comment by nicka** [comment]
@flomotlik Haha nice!!!

Maybe the other-way around with `--hide-events`? Because I feel new users(new to CloudFormation) should learn about the deployment process and how it works. I think it's even quite useful to know what you are actually "waiting" for.

**Comment by flomotlik** [comment]
>Maybe the other-way around with --hide-events? Because I feel new users(new to CloudFormation) should learn about the deployment process and how it works. I think it's even quite useful to know what you are actually "waiting" for.

I think is good for experienced users, but new users will be completely overwhelmed by a lot of noise that they know nothing about or understand (and don't really need either). e.g. a comment here: https://github.com/serverless/serverless/issues/1877#issuecomment-240405452 (which is a completely fine comment in terms of not being familiar with CF).

And any more experienced developer will use --hide-events all the time as you're not really interested in everything anyway and just want to see errors. Also in CI it just makes the log REALLY long which makes CI very hard to parse.

**Comment by nicka** [comment]
Alright makes sense. Is `--debug` already in use?

**Comment by flomotlik** [comment]
Don't think so, but I also wouldn't call it debug, as the expectation of `--debug` would be more than just printing CF events. Debug should show everything that happens inside of the serverless command imho (and for that we have the Debug environment variable). I like --all-events :)

**Comment by nicka** [comment]
Okidoki, I will start working on it and hopefully have something by the beginning of next week.

**Comment by nicka** [comment]
@serverless/team Making some progress. What do you think about the output so far?

<img width="791" alt="screen shot 2016-08-19 at 21 30 50" src="https://cloud.githubusercontent.com/assets/195404/17822371/1a0af19e-6657-11e6-8e18-232ceac95d15.png">
<img width="790" alt="screen shot 2016-08-19 at 21 49 41" src="https://cloud.githubusercontent.com/assets/195404/17822374/1ed3d11e-6657-11e6-8116-3b302b69a6d0.png">

**Comment by nicka** [comment]
@flomotlik Could you please add the `in-progress` label?

**Comment by flomotlik** [comment]
Done. This looks amazing. I want to use it instantly ☺️

**Comment by nicka** [comment]
@flomotlik Thanks!!! Still needs some refactoring so it can be re-used within createStack and updateStack. First wanted to check if everybody likes the new output ;)

**Comment by pmuens** [comment]
This looks really cool! Like it. Good job 🎉

**Comment by nicka** [comment]
@pmuens Thanks 🍻

Btw I decided to go with `--tail`, `-t` just like the `logs` plugin.

**Comment by mthenw** [comment]
@nicka great stuff there! I really like it.

Two things:
- What do you think about prefixing each line with "CloudFormation: ". So the printed log would look like:

```
Serverless: Checking stack update progress...
CloudFormation: UPDATE_IN_PROGRESS...
```

It would make clearer which lines are from framework and which are from CF. 

- I think that it should not be a part of standard output log. I would make that logs (and also currently printed logs) as a "info" level (https://github.com/serverless/serverless/issues/1720). So user can do `serverless deploy --log-level info` to see that logs. By default, assuming that command finished successfully, no logs should be printed. Or maybe only "Deployed" or something like that.

**Comment by ac360** [comment]
+1 to @mthenw's comment, with one design nitpik to consistently use hyphens instead of colons.  Not a big deal, but colons seem to be used heavily in the CF resource names, and don't want to add confusion.
+1 to @flomotlik's comment that we hide CF events by default (as long as we print out the full CF error at the end).  I recommend the common `--verbose` flag.  And trimming/refining the printed statements by Serverless so that they inform the user that CloudFormation is what's being used.

Examples:

```
sls deploy --verbose

Updating CloudFormation stack and reporting progress...
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - hello
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - hello2
CloudFormation - UPDATE_FAILED - AWS::Lambda::Function - hello2

   Serverless Error -------------------------------------
      An error occurred during deployment in CloudFormation:  [full error message]
```

```
sls deploy

Updating CloudFormation stack and reporting progress...

   Serverless Error -------------------------------------
      An error occurred during deployment in CloudFormation:  [full error message]
```

**Comment by flomotlik** [comment]
@ac360 in that case we would have a difference between the `Serverless:` and `Cloudformation -` output. Or did you mean removing the `Serverless:` part completely?

I'd be in favor of that (though I would do it in a separate PR).

`--tail` or `--verbose` They are both good, but I think I prefer verbose a little more. Not using `--tail` could send the signal we're not necessarily following along and not tailing the events or the deployment. So not using `--tail` could mean we're turning off some functionality, when the only thing we're turning off is the output. So I think `--verbose` is a little bit more appropriate here, though both are really close.

**Comment by nicka** [comment]
I'm going with `--verbose` quite common among `CLI's` thanks for pointing this out @ac360!

Regarding the actual output I will go with `Cloudformation -` 👍

**Comment by flomotlik** [comment]
Sounds great @nicka , looking forward to using it!
